api/src/service/domain/organization/user_permission_grant.spec.ts

```diff
@@ -0,0 +1,112 @@
+import { assert } from "chai";
+
+import { Ctx } from "../../../lib/ctx";
+import * as Result from "../../../result";
+import { NotAuthorized } from "../errors/not_authorized";
+import { ServiceUser } from "../organization/service_user";
+import { newUserFromEvent } from "./user_eventsourcing";
+import { grantUserPermission } from "./user_permission_grant";
+import { UserRecord } from "./user_record";
+
+const ctx: Ctx = { requestId: "", source: "test" };
+const root: ServiceUser = { id: "root", groups: [] };
+const alice: ServiceUser = { id: "alice", groups: ["alice_and_bob", "alice_and_bob_and_charlie"] };
+const bob: ServiceUser = { id: "bob", groups: ["alice_and_bob", "alice_and_bob_and_charlie"] };
+const charlie: ServiceUser = { id: "charlie", groups: ["alice_and_bob_and_charlie"] };
+
+const grantIntent = "user.intent.grantPermission";
+const userId = "dummy";
+const baseUser: UserRecord = {
+  id: userId,
+  createdAt: new Date().toISOString(),
+  organization: "dummy",
+  displayName: "dummy",
+  passwordHash: "password",
+  permissions: {},
+  address: "12345",
+  encryptedPrivKey: "encrypted",
+  log: [],
+  additionalData: {},
+};
+
+const baseRepository = {
+  getUser: async () => baseUser,
+};
+
+describe("Granting user permissions: permissions", () => {
+  it("Without the user.intent.grantPermission permission, a user cannot grant user permissions", async () => {
+    const result = await grantUserPermission(ctx, alice, userId, bob.id, grantIntent, {
+      ...baseRepository,
+    });
+
+    // NotAuthorized error due to the missing permissions:
+    assert.isTrue(Result.isErr(result), "Alice is not authorized to grant this permission");
+    assert.instanceOf(result, NotAuthorized, "The error is of the type 'Not Authorized'");
+  });
+
+  it("The root user can always grant user permissions", async () => {
+    const result = await grantUserPermission(ctx, root, userId, bob.id, grantIntent, {
+      ...baseRepository,
+    });
+
+    assert.isTrue(Result.isOk(result));
+  });
+});
+
+describe("Granting user permissions: updates", () => {
+  it("The permission is granted if the user has the correct permissions", async () => {
+    const result = await grantUserPermission(ctx, alice, userId, bob.id, grantIntent, {
+      getUser: async () =>
+        Promise.resolve({
+          ...baseUser,
+          permissions: { "user.intent.grantPermission": [alice.id] },
+        }),
+    });
+    if (Result.isErr(result)) {
+      throw result;
+    }
+
+    const sourcedUser = result.reduce(
+      (user, event) => newUserFromEvent(ctx, user, event),
+      baseUser,
```

Comment:
not baseUser but user that is used in `getUser`

Reply:
put user in separate variable `permissionTestUser` and used that one for sourcing